readme.md
ATTENTION: DO NOT USE THIS PACKAGE! Please use https://github.com/sindresorhus/file-type#readme


# file-type [![Build Status](https://travis-ci.org/sindresorhus/file-type.svg?branch=master)](https://travis-ci.org/sindresorhus/file-type)

> Detect the file type of a Buffer/Uint8Array

The file type is detected by checking the [magic number](http://en.wikipedia.org/wiki/Magic_number_(programming)#Magic_numbers_in_files) of the buffer.


## Install

```
$ npm install --save file-type
```


## Usage

### CLI
```
node fileInspector --file PATH_TO_FILE [--showSignature]
OR
node httpInspector --url URL_TO_FILE [--showSignature]
```
If you use optional parameter "showSignature" the signature is shown even if the file type is detected.


##### Node.js

```js
const readChunk = require('read-chunk'); // npm install read-chunk
const fileType = require('file-type');
const buffer = readChunk.sync('unicorn.png', 0, 262);

fileType(buffer);
//=> {ext: 'png', mime: 'image/png'}
```

or from a remote location:

```js
const http = require('http');
const fileType = require('file-type');
const url = 'http://assets-cdn.github.com/images/spinners/octocat-spinner-32.gif';

http.get(url, res => {
	res.once('data', chunk => {
		res.destroy();
		console.log(fileType(chunk));
		//=> {ext: 'gif', mime: 'image/gif'}
	});
});
```

##### Browser

```js
const xhr = new XMLHttpRequest();
xhr.open('GET', 'unicorn.png');
xhr.responseType = 'arraybuffer';

xhr.onload = () => {
	fileType(new Uint8Array(this.response));
	//=> {ext: 'png', mime: 'image/png'}
};

xhr.send();
```


## API

### fileType(buffer)

Returns an `Object` (or `null` when no match) with:

- `ext` - one of the [supported file types](#supported-file-types)
- `mime` - the [MIME type](http://en.wikipedia.org/wiki/Internet_media_type)

#### buffer

Type: `Buffer` `Uint8Array`

It only needs the first 262 bytes.


## Supported file types

- [`jpg`](https://en.wikipedia.org/wiki/JPEG)
- [`png`](https://en.wikipedia.org/wiki/Portable_Network_Graphics)
- [`gif`](https://en.wikipedia.org/wiki/GIF)
- [`webp`](https://en.wikipedia.org/wiki/WebP)
- [`cr2`](http://fileinfo.com/extension/cr2)
- [`tif`](https://en.wikipedia.org/wiki/Tagged_Image_File_Format)
- [`bmp`](https://en.wikipedia.org/wiki/BMP_file_format)
- [`jxr`](https://en.wikipedia.org/wiki/JPEG_XR)
- [`psd`](https://en.wikipedia.org/wiki/Adobe_Photoshop#File_format)
- [`zip`](https://en.wikipedia.org/wiki/Zip_(file_format))
- [`tar`](https://en.wikipedia.org/wiki/Tar_(computing)#File_format)
- [`rar`](https://en.wikipedia.org/wiki/RAR_(file_format))
- [`gz`](https://en.wikipedia.org/wiki/Gzip)
- [`bz2`](https://en.wikipedia.org/wiki/Bzip2)
- [`7z`](https://en.wikipedia.org/wiki/7z)
- [`dmg`](https://en.wikipedia.org/wiki/Apple_Disk_Image)
- [`mp4`](https://en.wikipedia.org/wiki/MPEG-4_Part_14#Filename_extensions)
- [`m4v`](https://en.wikipedia.org/wiki/M4V)
- [`mid`](https://en.wikipedia.org/wiki/MIDI)
- [`mkv`](https://en.wikipedia.org/wiki/Matroska)
- [`webm`](https://en.wikipedia.org/wiki/WebM)
- [`mov`](https://en.wikipedia.org/wiki/QuickTime_File_Format)
- [`avi`](https://en.wikipedia.org/wiki/Audio_Video_Interleave)
- [`wmv`](https://en.wikipedia.org/wiki/Windows_Media_Video)
- [`mpg`](https://en.wikipedia.org/wiki/MPEG-1)
- [`mp3`](https://en.wikipedia.org/wiki/MP3)
- [`m4a`](https://en.wikipedia.org/wiki/MPEG-4_Part_14#.MP4_versus_.M4A)
- [`ogg`](https://en.wikipedia.org/wiki/Ogg)
- [`opus`](https://en.wikipedia.org/wiki/Opus_(audio_format))
- [`flac`](https://en.wikipedia.org/wiki/FLAC)
- [`wav`](https://en.wikipedia.org/wiki/WAV)
- [`amr`](https://en.wikipedia.org/wiki/Adaptive_Multi-Rate_audio_codec)
- [`pdf`](https://en.wikipedia.org/wiki/Portable_Document_Format)
- [`epub`](https://en.wikipedia.org/wiki/EPUB)
- [`exe`](https://en.wikipedia.org/wiki/.exe)
- [`swf`](https://en.wikipedia.org/wiki/SWF)
- [`rtf`](https://en.wikipedia.org/wiki/Rich_Text_Format)
- [`woff`](https://en.wikipedia.org/wiki/Web_Open_Font_Format)
- [`woff2`](https://en.wikipedia.org/wiki/Web_Open_Font_Format)
- [`eot`](https://en.wikipedia.org/wiki/Embedded_OpenType)
- [`ttf`](https://en.wikipedia.org/wiki/TrueType)
- [`otf`](https://en.wikipedia.org/wiki/OpenType)
- [`ico`](https://en.wikipedia.org/wiki/ICO_(file_format))
- [`flv`](https://en.wikipedia.org/wiki/Flash_Video)
- [`ps`](https://en.wikipedia.org/wiki/Postscript)
- [`xz`](https://en.wikipedia.org/wiki/Xz)
- [`sqlite`](https://www.sqlite.org/fileformat2.html)
- [`nes`](http://fileinfo.com/extension/nes)
- [`crx`](https://developer.chrome.com/extensions/crx)
- [`xpi`](https://en.wikipedia.org/wiki/XPInstall)
- [`cab`](https://en.wikipedia.org/wiki/Cabinet_(file_format))
- [`deb`](https://en.wikipedia.org/wiki/Deb_(file_format))
- [`ar`](https://en.wikipedia.org/wiki/Ar_(Unix))
- [`rpm`](http://fileinfo.com/extension/rpm)
- [`Z`](http://fileinfo.com/extension/z)
- [`lz`](https://en.wikipedia.org/wiki/Lzip)
- [`msi`](https://en.wikipedia.org/wiki/Windows_Installer)

*SVG isn't included as it requires the whole file to be read, but you can get it [here](https://github.com/sindresorhus/is-svg).*

*PR welcome for additional commonly used file types.*

## Add new magic numbers
Determine the magic numbers by using the readChunk result to create a signature:
```
var signature = "";
for(var x = 0; x < 20; x++) {
    if (result[x]) signature += result[x].toString(16)+' ';
}
```

Use this signature to compare it with a list of magic numbers - see below for sources

## Test local files
Use the fileInspector.js to test local files. Please edit the file and add your fileName and then call:
```
node fileInspector.js

Result: 
DETECTED { ext: 'vtt', mime: 'text/plain' }
```

## Magic Number lists and databases
+ http://www.garykessler.net/library/file_sigs.html
+ http://tool.lu/magicbytes/
+ https://gist.github.com/navinpai/983632
+ http://lclevy.free.fr/raw/

## Related

- [file-type-cli](https://github.com/sindresorhus/file-type-cli) - CLI for this module


## License

MIT © [Sindre Sorhus](https://sindresorhus.com)


CHANGELOG.md
<a name="3.8.2"></a>

## [3.8.2](https://github.com/mmpro/file-type/compare/v3.8.1..v3.8.2) (2020-09-25 07:42:53)


### Bug Fix

* **App:** Fixed detecting Adobe Illustrator | MP | [c9f9a1bc46117ee89e94313ce0c3b22706d21b19](https://github.com/mmpro/file-type/commit/c9f9a1bc46117ee89e94313ce0c3b22706d21b19)    
Fixed detecting Adobe Illustrator  
Related issues: [undefined/undefined#AC-637](undefined/browse/AC-637)
### Chores

* **App:** Updated packages | MP | [5d078fdcb40e38eed12b4f430c7c0f826d3a473a](https://github.com/mmpro/file-type/commit/5d078fdcb40e38eed12b4f430c7c0f826d3a473a)    
Updated packages
<a name="3.8.1"></a>

## [3.8.1](https://github.com/mmpro/file-type/compare/v3.8.0..v3.8.1) (2020-06-30 11:27:33)


### Bug Fix

* **App:** Use mimetype image/heic for heic | MP | [ecead1c2e64b4bf0f2cb1456ba4fc6925818eb45](https://github.com/mmpro/file-type/commit/ecead1c2e64b4bf0f2cb1456ba4fc6925818eb45)    
Use mimetype image/heic for heic
### Style

* **App:** Lint fix | MP | [3a7b0a99472223c60ab331de5d3a8e808640f4b1](https://github.com/mmpro/file-type/commit/3a7b0a99472223c60ab331de5d3a8e808640f4b1)    
Lint fix
### Chores

* **App:** Prepare for ac-semantic-release | MP | [ec1061c57c1768d1125bb3dd4bef68ff80c5c361](https://github.com/mmpro/file-type/commit/ec1061c57c1768d1125bb3dd4bef68ff80c5c361)    
Prepare for ac-semantic-release
**3.8.21 - 02.02.2018**
+ updated: Detect ISOM HVC1 MP4 format 

**3.8.20 - 25.01.2018**
+ updated: Detect AIFC format 

**3.8.16 - 13.10.2017**
+ added: Detect EPS file (and distinguish from PS file)

**3.8.15 - 08.10.2017**
+ added: Zoom H6 projectfile (Heroglyph Project Format)

**3.8.14 - 21.09.2017**
+ added: RED Redcode RAW

**3.8.13 - 06.07.2017**
+ added: MPEG ADTS - related mmpro/ac-transcoding-server#142, mmpro/ac-transcoding-server#143, mmpro/ac-transcoding-server#144, mmpro/ac-transcoding-server#145
+ added: PXM file - related mmpro/ac-transcoding-server#136

**3.8.12 - 11.06.2017**
+ added: aac (MPEG-2 and MPEG-4)

**3.8.11 - 30.03.2017**
+ added: isom/mp4 compatible

**3.8.10 - 21.03.2017**
+ added: another MOV variant with AVCHD/Clip Wrap

**3.8.9 - 10.02.2017**
+ added: Adobe Indesign - 06 06 ed f5 d8 1d 46 e5 bd 31 ef e7 fe 74 b7 1d 44 4f 43 55

**3.8.8 - 15.01.2017**
+ updated: detect MPEG-TS if file header starts with 47 40

**3.8.7 - 15.01.2017**
+ added: JPEG2000 - 00 00 00 0C 6A 50 20 20 0D 0A 87 0A - related mmpro/ac-transcoding-server#107

**3.8.6 - 14.01.2017**
+ added: Apple QuickTime movie (unoptimized) - 00 00 00 01 6d 64 61 74 - mmpro/ac-transcoding-server#114

**3.8.5 - 13.01.2017**
+ added: additonal MPEG4 - 00 00 00 1c 66 74 79 70

**3.8.4 - 09.01.2017**
+ added: a different variation for SRT
+ added: detect SSA subtitle files

**3.8.3 - 08.01.2017**
+ added: magic numbers for MPEG TS

**3.8.2 - 08.01.2017**
+ added: magic numbers for aiff
+ added: httpInspector to quickly inspect a remote files

**3.8.1 - 27.09.2016**
+ added: magic numbers for mov (with byte orders wide and free), vim, srt and vtt
+ added: fileInspector to quickly inspect a local file
+ added: lodash for more convenience in fileInspector.js


test.js
import path from 'path';
import test from 'ava';
import readChunk from 'read-chunk';
import fileType from './';

function check(ext, name) {
	const file = path.join(__dirname, 'fixture', `${(name || 'fixture')}.${ext}`);
	const fileInfo = fileType(readChunk.sync(file, 0, 262)) || {};

	return fileInfo.ext;
}

const types = [
	'jpg',
	'png',
	'gif',
	'webp',
	'cr2',
//	'tif',
	'bmp',
	'jxr',
	'psd',
	'zip',
	'tar',
	'rar',
	'gz',
	'bz2',
	'7z',
	'dmg',
//	'mp4',
	'm4v',
	'mid',
	'mkv',
	'webm',
	'mov',
	'avi',
	'wmv',
	'mpg',
	'mp3',
	'm4a',
	'ogg',
	'opus',
	'flac',
	'wav',
	'amr',
	'pdf',
	'epub',
	'exe',
	'swf',
	'rtf',
//	'woff',
//	'woff2',
//	'eot',
	'ttf',
	'otf',
	'ico',
	'flv',
	'ps',
	'xz',
	'sqlite',
	'nes',
	'crx',
	'xpi',
	'cab',
	'deb',
	'ar',
	'rpm',
	'Z',
	'lz',
	'msi',
	'mxf'
];

const names = {
	woff2: ['fixture', 'fixture-otto'],
	woff: ['fixture', 'fixture-otto'],
	eot: ['fixture', 'fixture-0x20001'],
	mp4: ['fixture-imovie', 'fixture-isom', 'fixture-isomv2'],
	tif: ['fixture-big-endian', 'fixture-little-endian'],
	gz: ['fixture.tar'],
	xz: ['fixture.tar'],
	lz: ['fixture.tar'],
	Z: ['fixture.tar']
};

function testFile(type, name) {
	test(type, t => {
		t.is(check(type, name), type);
	});
}

types.forEach(type => {
	if ({}.hasOwnProperty.call(names, type)) {
		names[type].forEach(name => testFile(type, name));
	} else {
		testFile(type);
	}
});


httpInspector.js
const http = require('https');
const _ = require('lodash');
const argv = require('minimist')(process.argv.slice(2));
const fileType = require('./');

const url = _.get(argv, 'url')
let showSignature = _.get(argv, 'showSignature')
const signatureLength = _.get(argv, 'signatureLength', 272)
const startPos = _.get(argv, 'startPos')

http.get(url, res => {
   res.once('data', result => {
     res.destroy();

     const detectedFileType = fileType(result);

      // determine signature so we can compare it with magic number lists and add it to file-type
      let signature = "";
      for(let x = 0; x < signatureLength; x++) {
        if(!_.isNil(result[x])) signature += _.padStart(result[x].toString(16),2,0) + ' ';
      }

      if (!detectedFileType) {
        console.log("COULD NOT DETECT fileType for %s - here's the signature", url);
        showSignature = true
      }
      else {
        console.log("DETECTED", detectedFileType);
      }

      if (showSignature) {
        console.log(_.repeat('-', 3))
        console.log(signature)
        console.log(_.repeat('-', 3))
        console.log(result.toString().substr(0, signatureLength))
        console.log(_.repeat('-', 60))
      }
      if (startPos) {
        console.log('%s starts at position %s', startPos, signature.indexOf(startPos)/3)
      }
      

   })
})


fileInspector.js
const _ = require('lodash');
const readChunk = require('read-chunk');
const argv = require('minimist')(process.argv.slice(2));
const fileType = require('./');

const fileName = _.get(argv, 'file')
let showSignature = _.get(argv, 'showSignature')
const signatureLength = _.get(argv, 'signatureLength', 272)
const startPos = _.get(argv, 'startPos')

readChunk(fileName, 0, signatureLength)
  .then((result) =>{

    const detectedFileType = fileType(result);

    // determine signature so we can compare it with magic number lists and add it to file-type
    let signature = "";
    for(let x = 0; x < signatureLength; x++) {
      if(!_.isNil(result[x])) signature += _.padStart(result[x].toString(16),2,0) + ' ';
      //console.log(_.padStart(x,3), _.padStart(result[x].toString(16),2,0), result.toString().substring(x, x+1))
    }

    if (!detectedFileType) {
      console.log("COULD NOT DETECT fileType for %s - here's the signature", fileName);
      showSignature = true
    }
    else {
      console.log("DETECTED", detectedFileType);
    }

    if (showSignature) {
      console.log(_.repeat('-', 3))
      console.log(signature)
      console.log(_.repeat('-', 3))
      console.log(result.toString().substr(0, signatureLength))
      console.log(_.repeat('-', 60))
    }
    if (startPos) {
      console.log('%s starts at position %s', startPos, signature.indexOf(startPos)/3)
    }

  })
  .catch(function(err) {
    console.error('FAILED', err)
  });
